storage/v1alpha1/types_share.go

```diff
@@ -0,0 +1,149 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// SharedResource allows a backing Secret or ConfigMap to be shared across namespaces.
+// Pods can mount the shared Secret or ConfigMap by adding a CSI volume to the pod specification using the
+// "sharedresource.csi.storage.openshift.io" CSI driver and a reference to the SharedResource in the volume attributes:
+//
+// spec:
+//  volumes:
+//  - name: shared-secret
+//    csi:
+//      driver: sharedresource.csi.storage.openshift.io
+//      volumeAttributes:
+//        sharedResource: my-share
+//
+// For the mount to be successful, the pod's service account must be granted permission to get the named SharedResource object with an appropriate
+// ClusterRole and ClusterRoleBinding.
+//
+// apiVersion: rbac.authorization.k8s.io/v1
+// kind: ClusterRole
+// metadata:
+//   name: shared-resource-my-share
+// rules:
+//   - apiGroups:
+//       - storage.openshift.io
+//     resources:
+//       - sharedresources
+//     resourceNames:
+//       - my-share
+//     verbs:
+//       - get
```

Comment:
interesting.  So how does a regular user know which SharedResources are present so they can select one to include in their pod?

Reply:
@adambkaplan or @gabemontero would probably have a better answer to that question.

Reply:
the mental model has been they can only list the ones an admin has given them permission to access

so by default, a regular user does not have access to shares 

the admin must decide to create a permissive ClusterRole that allows list/get of all shares, regardless of share name

Reply:
How would a user with the ability to see the sharedresources in question be able to grant those rights to a serviceaccount?  And if the only way to get a list of all sharedresources is to have permission to use all sharedresources, is this subdivision achieving its goals?

Reply:
> How would a user with the ability to see the sharedresources in question be able to grant those rights to a serviceaccount? 

again "restrictive" mindset, albeit with more onus on the admin

the admin who gave the user the rights to see the shares is the one who grants the rights to the SA the user will use

if we are forcing to much work on the admin, or not providing enough power to the user, then presumably we need more levels of abstraction

conceivably something to sort out between tech preview and GA

> And if the only way to get a list of all sharedresources is to have permission to use all sharedresources, is this subdivision achieving its goals?

if a user could only see 3 of 8 shares, is it not reasonble to create a cluster role which has specific resourceName entries for those 3 shares, and then adds the get and list verbs?

will that not mean the list operation will only return the 3 the user has access to?

if the above understanding is flawed, then presumably we need more levels of abstraction ?

Reply:
Your API needs to be complete in terms of security and visiblity in alpha or the feedback you receive is not useful for determining whether or not you can graduate.

In a restrictive setting (which should be on-by-default in openshift), I don't see how a pod author locates SharedResources.  Granting GET permissions to a particular item does not allow LIST permission which show all content, not just ones you have GET access for.

Being able to describe a complete and secure story here is a blocker for alpha.

Reply:
OK, as we understand it then, the meets min is a "restricted LIST", which most likely mean new api object(s) that has the list of shares you can get.

Users are then given access to the new api objects that list the shares they have access to.

Or a more elegant abstraction than this that meets the goals.

There is also the question of roles/rolebindings  to attach to the pod SA.  Still not sure about that.  Or if we need to address for alpha.

Bottom line, we have to pause this effort/story and go back to the openshift EP for this.

We'll engage with you there @deads2k .... thanks.